src/auth/auth-provider.tsx

```diff
@@ -61,78 +60,40 @@ export function AuthProvider({
 } | null
 
-function useAuthentication(): [RefObject<AuthenticationPayload>, boolean] {
-  const initialValue = parseAuthCookie()
-  const authenticationPayload = useRef<AuthenticationPayload>(initialValue)
-  const pendingRefreshTokenPromise = useRef<Promise<void> | null>(null)
-
-  const [loggedIn, setLoggedIn] = useState(() => {
-    return initialValue !== null
-  })
-
-  function parseAuthCookie(): AuthenticationPayload {
-    try {
-      const cookies = typeof window === 'undefined' ? {} : Cookies.get()
-
-      const { access_token, id_token } = JSON.parse(
-        cookies['auth-token']
-      ) as Token
-
-      const decoded = jwt_decode<{
-        username: string
-        id: number
-      }>(id_token as string)
-
-      return {
-        username: decoded.username,
-        id: decoded.id,
-        token: access_token as string,
-        refreshToken,
-        clearToken,
-      }
-    } catch {
-      return null
+function parseAuthCookie(session: Session): AuthenticationPayload {
+  try {
+    const username = (session?.identity?.traits as { username: string })
+      ?.username
+    const legacyId = (
+      session?.identity?.metadata_public as { legacy_id: number }
+    )?.legacy_id
+    return {
+      username,
+      id: legacyId,
     }
+  } catch {
+    return null
   }
+}
 
-  // eslint-disable-next-line @typescript-eslint/require-await
-  async function refreshToken(usedToken: string) {
-    async function startRefreshTokenPromise(): Promise<void> {
-      if (typeof window === 'undefined') return
-
-      await fetch('/api/auth/refresh-token', {
-        method: 'POST',
-      })
-
-      authenticationPayload.current = parseAuthCookie()
-      pendingRefreshTokenPromise.current = null
-    }
-
-    const currentCookieValue = parseAuthCookie()
-    if (currentCookieValue === null || currentCookieValue.token !== usedToken) {
-      // Cooke has a newer token than the one we used for the request. So use that instead.
-      authenticationPayload.current = currentCookieValue
-      return
-    }
-
-    if (!pendingRefreshTokenPromise.current) {
-      // Only initiate the token refresh request if it has not been started already.
-      pendingRefreshTokenPromise.current = startRefreshTokenPromise()
-    }
+function useAuthentication(): [RefObject<AuthenticationPayload>, boolean] {
+  const [session, setSession] = useState<Session | null>(null)
+  const authenticationPayload = useRef<AuthenticationPayload>(null)
 
-    return pendingRefreshTokenPromise.current
-  }
+  useEffect(() => {
+    void (async () => {
+      const { data } = await kratos.toSession().catch()
+      setSession(data)
+      authenticationPayload.current = parseAuthCookie(data)
+      setLoggedIn(data !== null)
+    })()
+  }, [])
 
-  function clearToken() {
-    if (!loggedIn) return
-    Cookies.remove('auth-token')
-    setLoggedIn(false)
-  }
+  const [loggedIn, setLoggedIn] = useState(() => {
+    return session !== null
+  })
```

Comment:
I'd like to change that up so the information whether the user is logged in is synchronous. Currently, it's always false until kratos reports info about the session. Which isn't necessary, since we already know that the user is most likely logged in if the cookie exists (if the session is invalid, we might need to clean that cookie manually if kratos doesn't do that for us, but yeah). Similarly: we could think about storing the session information locally, too, so that we don't need to wait for the response. And only check if the session is still valid or something.

Reply:
Okay, one slight problem: we can't check the cookie directly since it's httpOnly. But what we can do is the following:

- Persist the response of `toSession` to a cookie so that it's existent on first render.
- Use that cookie as initialValue to detect the user as logged in "optimistically".
- Still do the `toSession` to check if the session is still valid. If it's not (401 status code), then remove the cookie and report the user as not logged in anymore.

Reply:
And bonus points if we also set that session cookie directly after log in so that we never have a wrong `null` in the beginning (which would lead to flashing content).

Reply:
Please take a look at https://github.com/serlo/frontend/pull/1819/commits/bb16c8a431f77883d039f4e7ac8b8582aec46142 and see if that is what you mean

Reply:
Added two comments to https://github.com/serlo/frontend/commit/bb16c8a431f77883d039f4e7ac8b8582aec46142 which aren't shown in the PR directly for some reason.

Reply:
Alright, I answered you comments there.

If you think it is overall good, I will just change it instead of persisting the whole response of the toSession, I will store a simple value like 1 or 'true' (and of course refactor it). The whole response is too much info exposed (like user email) that would be available to malicious sites.

Reply:
Malicious sites don't have access to the cookie, though? But yeah, you can remove many parts, but you probably need the legacy id, right?

Reply:
Sorry,  you are right, I was afraid of the possibility of doing it inside an iframe, until I read it is indeed not possible.

But anyway, I get now I understood you wrong at the first time and you mean with initialValue also the data about the user (not only that she is logged in).

Reply:
done (I hope) in [22fc88c](https://github.com/serlo/frontend/pull/1819/commits/22fc88c88275cd152f85ce5e24a6b9d434671149)